### PR TITLE
FT MOTION: add individual axis shaping and new buffer management

### PR DESCRIPTION
**Pull request by narno2202**
### Description

@ulendoalex update the Marlincollaborative repository with a new buffer system management and the possibility to set shaping per axis. This PR adds individual axis shaping function, adapt the FTMotion menu accordingly and the new buffer managemnt system. This PR is a mix between #26720 last working code (homing works for coreXY as reported in anotther discussion, BLTOUCH HS Mode...) and @ulendoalex code with some minor changes. This PR suspercedes #26720 which needs more work to be fully functional. 

- FTMotion can be set on or off globally (only the motion system not the shapers).
- Shaping can be set to any shaper (ZV, ZVD ...) or disabled  on an individual axis basis (only for X and Y axis).
- Buffer management system is left untouched

### Requirements

Ideally, this PR should be merged before the other `FT_MOTION` PRs :
- #26785
- #26720
- #26670 

### Configurations

Just enable `FT_MOTION` in `Configuration_adv.h`

### Related Issues

None

### CODE REVIEW

**Comment by ulendoalex** [comment]
@narno2202 

Thanks for creating this, I intended to issue the pull request myself after more testing. I'm fairly confident in the changes, but please tag me if you encounter any issues.

**Comment by narno2202** [comment]
This is not perhaps the best solution, but I was thinking of progressively incorporate your code in current Marlin bugfix after testing and debugging the branch in your repository. 
`BLTouch HS mode` is broken and needs `axis_did_move.reset()` to work (it's  in my PR in Marlincollaborative) and again if `FTM_BATCH_SIZE` is not equal to `FTM_WINDOW_SIZE / 2` : motion is broken (this time for `FTM_BATCH_SIZE <  FTM_WINDOW_SIZE / 2`)

**Comment by narno2202** [comment]
@ulendoalex , in `reset()` function `makeVector_batchIdx = TERN(FTM_UNIFIED_BWS, 0, _MIN(BATCH_SIDX_IN_WINDOW, FTM_BATCH_SIZE));` instead of `makeVector_batchIdx = BATCH_SIDX_IN_WINDOW;` solves the bug with asymetrical window and batch size (ratio not equal to 2).

**Comment by bmourit** [comment]
This throws a warning about a possible uninitialized variable in menu_motion.cpp line 461, 466 and 470. I am pretty sure it is safe to ignore, but maybe you want to avoid?

Adding a "default:" to the switch cases fixes this warning.

**Comment by narno2202** [comment]
Strange, there is no switch .. case here. I do a compilation test for the simulator with the menu enabled and there is no warning.

**Comment by bmourit** [comment]
in motion_menu.cpp switches at line 414, 420 and 434.

**Comment by narno2202** [comment]
Still no issue.

**Comment by bmourit** [comment]
VS Code is showing it in the console during a real build.

**Comment by narno2202** [comment]
You're right, corrected

**Comment by bmourit** [comment]
Awesome! I'm excited to see this addition (once I figure my machine's issue out). Great work!

**Comment by ulendoalex** [comment]
> @ulendoalex , in `reset()` function `makeVector_batchIdx = TERN(FTM_UNIFIED_BWS, 0, _MIN(BATCH_SIDX_IN_WINDOW, FTM_BATCH_SIZE));` instead of `makeVector_batchIdx = BATCH_SIDX_IN_WINDOW;` solves the bug with asymetrical window and batch size (ratio not equal to 2).

@narno2202 
Thanks for this. I haven't gotten around to investigating this bug yet; hopefully this week.

FYI related to this PR, I did find some bugs and have updated our repository accordingly. You may wish to take a look at those changes to make sure you take them (or already fixed them).

**Comment by thisiskeithb** [comment]
@narno2202: Can you merge in the latest `bugfix-2.1.x` & resolve the current file conflicts?

It’d be good to get FT Motion back into working order before `2.1.3` is released.

**Comment by thinkyhead** [comment]
Resolving conflicts is no fun, so I went ahead and did that. Be sure to check my work in case something was missed. This code overlaps a little with recent planner / stepper changes, so we want to be sure not to lose those here.

The documentation for `M493` in the MarlinDocumentation repo and on the RepRap Wikl will need to be updated to reflect these changes, and we should make some effort to settle on a final specification for the `M493` G-code that we would want to advocate for other firmwares to adopt as well.

Is it commonly needed to allow X and Y (and I presume other axes in future) to each use a different form of shaping? I can see how it may be useful, but perhaps it should be optional. A base implementation that only stores and implements the parameters for a single form of shaping will reduce memory use and code complexity a little for those who don't need this option, and every but helps in the motion code.

File: Marlin/src/gcode/feature/ft_motion/M493.cpp [hunk 1]
```diff
@@ -28,38 +28,48 @@
...
-      }
-      SERIAL_ECHOPGM(" shaping");
+    if (CMPNSTR_HAS_SHAPER(X_AXIS)) {
+      SERIAL_ECHOPGM(" with X/A");
+      say_shaper_type(ftMotion.cfg.cmpnstr[X_AXIS]);
+    }
+  #endif
+  #if HAS_Y_AXIS
+    if (CMPNSTR_HAS_SHAPER(Y_AXIS)) {
+      SERIAL_ECHOPGM(" and with Y/B");
```
Comment (thinkyhead) [inline Marlin/src/gcode/feature/ft_motion/M493.cpp]: We can say "with X shaper" / "with Y shaper" on standard Cartesian, and just print "with A shaper" / "with B shaper" on Core XY. That should make it a little clearer for both types of setups.

**Comment by thisiskeithb** [comment]
> Is it commonly needed to allow X and Y (and I presume other axes in future) to each use a different form of shaping?

I’m pretty sure Klipper allows it, but RRF doesn’t and they [explain why in their docs](https://docs.duet3d.com/User_manual/Tuning/Input_shaping#why-we-dont-support-different-shaping-for-the-x-and-y-axes):

> ### Why we don't support different shaping for the X and Y axes
>
> Two reasons:
>
> 1. It's not a good idea, for at least two reasons:
>   - Supporting different input shaping frequencies and/or algorithms on the X and Y axes means that if input shaping is applied to printing moves other than pure X or pure Y moves, the tool path no longer follows the path commanded by the slicer during acceleration and deceleration. As a result, artefacts will occur in the print, especially around corners. The artefacts can be reduced by a smoothing process (as is done by Klipper firmware), but of course this reduces detail, for example by rounding corners that are supposed to be square. The best compromise between reducing artefacts caused by input shaping and reducing detail caused by smoothing may depend on the print.
>
>   - Because input shaping modifies the acceleration profile of the move, when input shaping is applied to moves with extrusion the same modified acceleration profile should be applied to the extruder (and further modified by pressure advabnce, if it is configured) so that the extrusion rate stays in step with the movement. RepRapFirmware 3.5beta3 and later sychrhonise extruder acceleration with movement acceleration in this way. If different input shaping is applied to X and Y axes, extrusion can't be synchronised to both, so the extrusion won't be uniform.
>
> 2. It's rarely if ever necessary to use different shaping on X and Y axes, because a single input shaper can suppress a wide range of resonant frequencies. The EI3 shaper suppresses ringing over a frequency range of 3:1. Resonant frequencies vary in proportion to the **square root** of the mass being moved, and the **square root** of the compliance of the belt or other elastic medium. Therefore you would need these elements to differ by a factor of more than 9:1 between the X and Y axes for the resonant frequencies to exceed a 3:1 ratio. In practice, the resonant frequencies to be suppressed rarely exceed a ratio of 2:1.

**Comment by narno2202** [comment]
Fix an error in SanityCheck  and motions are fine, homing is ok (sensorless homing), BLTouch HS mode is working (Cartesian printer with 2 Y steppers, 3 Z steppers and 1 X stepper, TMC2226 and BLTouch).

**Comment by vovodroid** [comment]
Is PR ready for test?

**Comment by narno2202** [comment]
For me yes, this is the PR currently runing on my printer with some changes for my screen (MKS H43).

**Comment by vovodroid** [comment]
Well, it's somehow alive now, but still doesn't work properly. In simplest scenario - build with both IS and FTM, disable them in menu, store settings, restart, home - both X and Y move simultaneously until one them reach limit and motor starts skipping ( TMC2209 sensorless homing).

If before homing enable FTM, manually move a bit X and Y, and then home - X and Y home, next Z goes down a bit with very slow speed and then Marlin stops with error "Homing failed".

Printer is Cartesian, BTT SKR3, TMC2209.

**Comment by narno2202** [comment]
When I used FT_MOTION for the first time, I needed to change TMC sensitivity values to get a fully working sensorless homing. I follow your description, but I have no error.

**Comment by vovodroid** [comment]
> I needed to change TMC sensitivity values to get a fully working sensorless homing

Even if it's disabled in menu?

**Comment by vovodroid** [comment]
I've built with 

```
#define FT_MOTION
//#define INPUT_SHAPING_X
//#define INPUT_SHAPING_Y
```
and without enabling set sensitivity to the highest value 255 for TMC2209. Bump to stopper still is not detected.

**Comment by vovodroid** [comment]
I tried according to @grizzleeadam [comment](https://github.com/MarlinFirmware/Marlin/issues/27076#issuecomment-2119547647)

> FT Motion seems to be working fine for me - so long as Adaptive_Step_Smoothing and Edge_Stepping are disabled

It didn't help, axes still crash to stops (with FTM disabled in menu).

**Comment by grizzleeadam** [comment]
Other options I have disabled:
-S_Curve (haven’t tested with FT_Motion yet)
-Linear Advance (I’m using the new version included with FT_Motion)
-Input Shaping (just to make sure it isn’t conflicting with FT_Motion)

I could share my config files if it would help. Curious to see what other
configurations cause issues with FT_Motion or if it’s inherently bugged.

**Comment by vovodroid** [comment]
New test, build with

```
#define FT_MOTION
#define FTM_DEFAULT_MODE        ftMotionMode_ENABLED 
#define ADAPTIVE_STEP_SMOOTHING (doesn't matter)
//#define INPUT_SHAPING_X
//#define INPUT_SHAPING_Y
```
and adjusted sensorless homing sensitivity.
After restart axes X/Y could home, but Z descents for a while and next Marlin stops with homing error. But if I manually press the probe it succeeded.

Here is video and attached config files with minimal changes.



https://github.com/MarlinFirmware/Marlin/assets/8691781/9e67b96b-4358-470f-a449-da58bd2b480e




[Marlin.zip](https://github.com/MarlinFirmware/Marlin/files/15375486/Marlin.zip)

**Comment by vovodroid** [comment]
Also XY homing fails if FTM disabled in menu.

**Comment by narno2202** [comment]
I apologize, @vovodroid  you're right. Could you please test my last commit before @thinkyhead  merge last marlin code https://github.com/MarlinFirmware/Marlin/pull/26848/commits/e03b0bcbddba52c5cf7f948bf45f904a2efb2e1f. I think the culprit is in the endstops code introduced in the previous FT_MOTION PR. I'm still waiting for an answer about : is endstops managed by Temperature ISR is enough once the printer has been homed.

**Comment by vovodroid** [comment]
Test https://github.com/MarlinFirmware/Marlin/commit/e03b0bcbddba52c5cf7f948bf45f904a2efb2e1f or previous one?

**Comment by narno2202** [comment]
@vovodroid , please test https://github.com/MarlinFirmware/Marlin/commit/e03b0bcbddba52c5cf7f948bf45f904a2efb2e1f .

**Comment by vovodroid** [comment]
Unfortunately result is the same - XY homed, goes to center, Z moves a bit up, probe is deployed, Z moves up a bit more, then goes down until Marlin stops.

**Comment by narno2202** [comment]
@vovodroid, the main difference between your Configuration.h file and mine is ENDSTOSPULLUPS which is enabled in my config (I can't compile Marlin with only #define ENDSTOPPULLUP_ZMIN if I comment ENDSTOPULLUPS). After @thinkyhead changes, I also have an uncommon Z motion after Z homing which not present before.

**Comment by thinkyhead** [comment]
You rang? How may I assist?

**Comment by thinkyhead** [comment]
Since setting the shaper parameter is equivalent to turning it on/off for an axis, we might as well keep the ability to set them to different shapers as standard. But we might consider providing some UI to allow users to quickly toggle the shaper on and off for all axes while still keeping ft motion as the active motion system.

**Comment by thinkyhead** [comment]
If FBS is not going to be a future value for `ftMotionMode_t` then we might as well drop the `ftMotionMode_t` type and just use a `bool` simply indicating that FT Motion is enabled or not.

**Comment by vovodroid** [comment]
May I test commit [e0c0a69](https://github.com/MarlinFirmware/Marlin/pull/26848/commits/e0c0a690e6d1ae37b401ec833699f27d707dc7c9)?

**Comment by thinkyhead** [comment]
Should be in a more testable state now.

**Comment by vovodroid** [comment]
The same, besides that I needed to reduce sensitivity more close to original values.

**Comment by narno2202** [comment]
Before the new endstops management system, sensorless homing was fine with or without FT_MOTION once It was tuned. Now with the same settings, sensorless homing is OK with FT_MOTION and abort with standard motion (same behavior in current bugfix).

**Comment by vovodroid** [comment]
> sensorless homing is OK with FT_MOTION

But there is Z axis problem in my printer.  I can take logs via USB port, just tell me what enable in firmware.

**Comment by vovodroid** [comment]
Commit [831ced4](https://github.com/MarlinFirmware/Marlin/pull/26848/commits/831ced42712aba1f5a08004e5444911e27464207) - the same.

**Comment by grizzleeadam** [comment]
You mentioned sensor less homing before. Are you using end stops on X and Y
or not? Does everything work properly with FT_Motion commented out, and/or
with it enabled and turned on in the menu?

When you’re having the issue, do X and Y move normally and it’s just the Z
issue?

**Comment by vovodroid** [comment]
> Are you using end stops on X and Y or not?

I have no physical switches, but TMC2209 DIAG pins are connected to corresponded board connectors.

> Does everything work properly with FT_Motion commented out

Stop isn't triggered, but probably it's sensitivity matter, I'll test it again.

> just the Z issue?

Yes.

**Comment by vovodroid** [comment]
> Stop isn't triggered, but probably it's sensitivity matter, I'll test it again.

Sorry, I confused here - with FTM commented out all works properly.

**Comment by narno2202** [comment]
I think it's "hardware related" for your Z axis as you use a Biqu Microprobe.All my 3 probes work with FTM enabled ( a BLTouch, a Creality CR Touch and another BLTouch clone). Now my only failure is sensorless homing when FTM is disabled and , it's related to the new endstops management introduced in last bugfix. 
@vovodroid , did a FTM version ever work and which one?

**Comment by vovodroid** [comment]
I can't even imagine how FTM  could be affected by Microprobe.
FTM never worked for me. First axes didn't move, now Z home failes. I'll collect debug output for homing vis USB and update.

**Comment by narno2202** [comment]
I mean the opposite : Microprobe affected by FTM but not other probes so "hardware realted". It's strange that no FTM version does not work with your printer as I have a cartesian printer too not an Ender 3 style but an Ender 5 style and it works for a long time now

**Comment by vovodroid** [comment]
I don't think Microprobe itself afected. It's nor false neither no triggering at all. I'll collect logs and let's see what happens.

**Comment by narno2202** [comment]
@vovodroid ; looking deeper in your Configuration.h file, your Microprobe settings are not those recommended on Bigtreetech github for a SKR3 motherboard. Have you test them?

**Comment by thisiskeithb** [comment]
> your Microprobe settings are not those recommended on Bigtreetech github for a SKR3 motherboard

I added native support so users only have to enable a couple defines. I also added sanity checks around each MicroProbe version and fixed a bug where the probe was not considered a “stowable probe” due to how BIQU recommended users enable it in their docs.

BIQU will update their docs to match once we have a tagged release of Marlin with MicroProbe support (hopefully `2.1.3`).

**Comment by vovodroid** [comment]
> your Microprobe settings are not those recommended on Bigtreetech github for a SKR3 motherboard. Have you test them?

As Keith mentioned BIQU doc is not correct. I successfully use Microprobe all the time.

**Comment by thisiskeithb** [comment]
> I successfully use Microprobe all the time.

No issues here either.

> [!NOTE]  
> MicroProbe != BLTouch. They do not work the same. `PROBE_ENABLE_DISABLE` is a pretty basic "probe" type since it only provides a triggered signal when enabled.

I have a couple machines here that use `PROBE_ENABLE_DISABLE` (Hurakan with MicroProbe and B1 SE Plus with a strain gauge hotend), so I can run some basic tests this weekend with & without sensorless homing and report back.

**Comment by bmourit** [comment]
Just confirming that the newest code works best so far for real endstop triggers (not sensorless homing). Both with and without endstop interrupt feature enabled on an Ender3. If I manually move any axis first, and then home, I get an error as previously reported. But so far this is the only version that doesn't grind my motors. All the other versions ignored the endsstops and tried to move anyway. And I have closely tested each change.

**Comment by vovodroid** [comment]
Well, I found that Z homing fails inside 

```
#if ENABLED(VALIDATE_HOMING_ENDSTOPS)
  // If the last move failed to trigger an endstop, call kill
  void Endstops::validate_homing_move() {
    if (trigger_state()) hit_on_purpose();
    else kill(GET_TEXT_F(MSG_KILL_HOMING_FAILED));
  }
#endif
```

So I disabled `VALIDATE_HOMING_ENDSTOPS`, enabled logs and try to home. There was preliminary triggering in the air, probably that's the reason why `validate_homing_move` kills Marlin. Here is Z homing log without `VALIDATE_HOMING_ENDSTOPS`, I hope it helps.

```
>>> G28  X-10.00 Y0.00 Z10.00
Machine Type: Cartesian
Probe: BIQU_MICROPROBE_V2
Probe Offset X10.00 Y10.00 Z0.00 (Right-Back & Same Z as Nozzle)
Y current: 800 -> 100
remember_feedrate_scaling_off: fr=66.67 100%
Raise Z before homing:
do_z_clearance(15.00 [10.00 to 15.00], 0)
do_blocking_move_to_z(15.00, 4.00)
>>> do_blocking_move_to  X-10.00 Y0.00 Z10.00
>  X-10.00 Y0.00 Z15.00
<<< do_blocking_move_to  X-10.00 Y0.00 Z15.00
>>> home_z_safely  X-10.00 Y0.00 Z15.00
  current_position= X-10.00 Y0.00 Z15.00 : sync_plan_position
  destination= X112.00 Y206.00 Z15.00 : home_z_safely
do_blocking_move_to_xy(112.00, 206.00, 0.00)
>>> do_blocking_move_to  X-10.00 Y0.00 Z15.00
>  X112.00 Y206.00 Z15.00
echo:busy: processing
<<< do_blocking_move_to  X112.00 Y206.00 Z15.00
>>> homeaxis(Z)
  current_position= X112.00 Y206.00 Z15.00 : Probe::set_deployed
deploy=1 no_return=0
Raise Z to 10.00
do_z_clearance(10.00 [15.00 to 10.00], 0)
>>> Probe::probe_specific_action  X112.00 Y206.00 Z15.00
<<< Probe::probe_specific_action  X112.00 Y206.00 Z15.00
>>> do_blocking_move_to  X112.00 Y206.00 Z15.00
>  X112.00 Y206.00 Z15.00
<<< do_blocking_move_to  X112.00 Y206.00 Z15.00
Home Fast: -360.00mm
>>> do_homing_move  X112.00 Y206.00 Z15.00
...(Z, -360.00, [4.00])
<<< do_homing_move  X112.00 Y206.00 Z15.00
Move Away: 5.00mm
>>> do_homing_move  X112.00 Y206.00 Z15.00
...(Z, 5.00, 4.00)
echo:busy: processing
<<< do_homing_move  X112.00 Y206.00 Z15.00
Re-bump: -10.00mm
>>> do_homing_move  X112.00 Y206.00 Z15.00
...(Z, -10.00, 2.00)
echo:busy: processing
echo:busy: processing
echo:busy: processing
<<< do_homing_move  X112.00 Y206.00 Z15.00
>>> set_axis_is_at_home(Z)
*** Z homed with PROBE (Z_MIN_PROBE_USES_Z_MIN_ENDSTOP_PIN) ***
> (M851 Z0.00)
> home_offset[Z] = 0.00
  current_position= X112.00 Y206.00 Z0.00 : 
<<< set_axis_is_at_home(Z)
  current_position= X112.00 Y206.00 Z0.00 : sync_plan_position
  current_position= X112.00 Y206.00 Z0.00 : > AFTER set_axis_is_at_home
  current_position= X112.00 Y206.00 Z0.00 : Probe::set_deployed
deploy=0 no_return=0
Raise Z to 10.00
do_z_clearance(10.00 [0.00 to 10.00], 0)
do_blocking_move_to_z(10.00, 4.00)
>>> do_blocking_move_to  X112.00 Y206.00 Z0.00
>  X112.00 Y206.00 Z10.00
echo:busy: processing
<<< do_blocking_move_to  X112.00 Y206.00 Z10.00
>>> Probe::probe_specific_action  X112.00 Y206.00 Z10.00
<<< Probe::probe_specific_action  X112.00 Y206.00 Z10.00
>>> do_blocking_move_to  X112.00 Y206.0

>  X112.00 Y206.00 Z10.00
<<< do_blocking_move_to  X112.00 Y206.00 Z10.00
<<< homeaxis(Z)
<<< home_z_safely  X112.00 Y206.00 Z10.00
  current_position= X112.00 Y206.00 Z10.00 : sync_plan_position
Restore driver current...
>>> do_move_after_z_homing  X112.00 Y206.00 Z10.00
do_z_clearance(5.00 [10.00 to 5.00], 1)
do_blocking_move_to_z(5.00, 4.00)
>>> do_blocking_move_to  X112.00 Y206.00 Z10.00
>  X112.00 Y206.00 Z5.00
echo:busy: processing
<<< do_blocking_move_to  X112.00 Y206.00 Z5.00
<<< do_move_after_z_homing  X112.00 Y206.00 Z5.00
restore_feedrate_and_scaling: fr=66.67 100%
X:112.00 Y:206.00 Z:5.00 E:0.00 Count X:8960 Y:16480 Z:2000
<<< G28  X112.00 Y206.00 Z5.00
```

**Comment by vovodroid** [comment]
Did I miss any option preventing Z going down below current zero without limit?

**Comment by vovodroid** [comment]
Here is compare report for my test and working config files (FTM on the left, working on the right). I don't see anything explaining Z homing fail.

[3D.zip](https://github.com/MarlinFirmware/Marlin/files/15443240/3D.zip)

**Comment by thinkyhead** [comment]
> Did I miss any option preventing Z going down below current zero without limit?

That is normal for external (`G0`, `G1`, …) moves, but internal moves (like those used in homing) are not limited.

> probably that's the reason why `validate_homing_move` kills Marlin

This function specifically checks whether the Z endstop was triggered by the last move. If it wasn't it goes to KILL, on the assumption that the endstop was never hit, for some reason.

The routines that register the endstop states depend on knowing that an axis is moving, and that it is moving in the direction towards the endstop. If the flag indicating motion is not being set correctly, or the flag indicating the axis current direction is not being set correctly, then the endstop hit will not be properly registered. This should be pretty obvious because the axis will grind (possibly damaging something).

We were aiming on the last go-round to make sure we set these global states in a way that persists for the duration of the linear move (i.e., not just during a single step) so that endstops are continually monitored. We should do another round of checks on the motion and direction states to make sure they are being set in the useful way as described. If we determine that they are okay, then we'll have to look for other factors that could contribute to this issue.

**Comment by thinkyhead** [comment]
I want to merge this, but first I want to install and test it on a machine and do various things to try to make it break. Maybe I can do some of that testing in the Sim. Anyway, once I have given it a thorough run-through and applied any needed fixes then we'll be ready to go forward with the next round of adjustments.

I have marked PRs that we need to complete soon with a "Version 2.1.3" Milestone, and those that are held for later with an "After 2.1.3" Milestone. This will include all items related to FT Motion, TMC2240, and a few others. It would be great to get these things all tested and fixed up before the solstice.

**Comment by thinkyhead** [comment]
I've fixed the homing issue introduced in PR #26720 where all homing was broken if `FT_MOTION` was enabled in the config. I tested homing (sensorless) using the current state of `bugfix-2.1.x` with FT Motion enabled. Homing failed. I'll check again using the code in this PR, and if it still fails I'll dive deeper into the cause. In the long run it will be good to be able to use FT Motion exclusively, so the sooner we get endstop moves working, the better.

**Comment by narno2202** [comment]
On my side sensorlees homing is broken since introduction  of the new endstops update function with millis function. Perhaps this should be reverted if standard homing is solved with your last commit. I will do some tests.

**Comment by thisiskeithb** [comment]
> I've fixed the homing issue introduced in PR https://github.com/MarlinFirmware/Marlin/pull/26720 where all homing was broken if FT_MOTION was enabled in the config. I tested homing (sensorless) using the current state of bugfix-2.1.x with FT Motion enabled. Homing failed.

Can confirm that sensorless homing is broken if `FT_MOTION` is compiled in + enabled on my BIQU BX with an SKR SE BX V2.0 (STM32H7) with TMC2226s (TMC2209s) and custom Prusa Bear with a BTT002 V1.0 (STM32F4) & TMC5160s. If `FT_MOTION` is compiled in but disabled, homing works as it should.

For my Bear with 5160s, X will either attempt to home & grind or it won't move at all and can be easily moved by hand until the printer halts due to a homing failed error. The Y axis will sometimes home fine or it will immediately trigger/show homed.

For my BX, both X & Y grind until a power cycle or I let it go for an uncomfortably long time. After letting it grind for ~15 seconds, it'll show homed. I don't know why it's not failing/halting before then.

Also, these machines sound *terrible* with `FT_MOTION` enabled, so there's probably a lot of re-tuning required compared to standard Input Shaping apparently. 😄

**Comment by narno2202** [comment]
I confirm that with removal of  

```
millis_t move_end_ti = millis() + SEC_TO_MS(FTM_TS*(float)(max_intervals + num_samples_cmpnstr_settle() + (PROP_BATCHES+1)*FTM_BATCH_SIZE) + ((float)FTM_STEPPERCMD_BUFF_SIZE/(float)FTM_STEPPER_FS));

  #if CORE_IS_XY ...
``` 
in ft_motion.cpp and of 
 ```
#if ENABLED(FT_MOTION)
      if (ftMotion.cfg.active)
        return (neg = ftMotion.axis_moving_neg(head)) || ftMotion.axis_moving_pos(head);
    #endif
```
in endstops.cpp, sensorless homing is fine . I can't test standard homing as my printer doesn't have mechanical endstops. Z homing with probe is also fine (was already fine).

**Comment by thinkyhead** [comment]
> I confirm that with removal of…

I went ahead and reverted to the previous method in `endstops.cpp` to check the motion state with `stepper.axis_is_moving(…)` and direction with `stepper.motor_direction(…)`.

For these flags to be reliable, all we really need is for them to be concurrent with the current motion and to persist for the duration of the segment. So, the global `axis_did_move` flags just need to be set when the movement block is started and cleared after the movement block is done.

Theoretically, FT Motion could have already fetched and queued up DIR/STEP commands for additional blocks while still processing a previous block, so we are required to propagate FT Motion's "internal" move/dir states to the global flags for the benefit of endstops. Using the global DIR state for FT Motion shouldn't be a problem because it changes infrequently, so we just need to make sure it's only being set by FT Motion commands when FT Motion is handling motion. The global `axis_did_move` flags should be set (OR'ed) as soon as a STEP command arrives for an axis, and cleared as soon as we know that a block's steps have been completed. Previously I created a special FT command to mark the end of a block, but that 9th bit caused the command size to double to 16 bits. I don't think it's possible to synthesize some "impossible" command combination to use instead, because theoretically any DIR/STEP command combo from 0x00-0xFF should be legal.

So, I'll compile all this into the Simulator and play around until I have a reliable solution.

**Comment by thinkyhead** [comment]
> …these machines sound terrible with `FT_MOTION` enabled…

Basic movement should be very nice with FT Motion, even if parameters aren't tuned perfectly, so I suspect a bug fell in. I'll be sure to look closely at the step timing in the simulator. It might be a good idea to loop some more smart cookies into this one, such as @mh-dm….

**Comment by narno2202** [comment]
Looking at the code, I realize that markBlockStart was always false in ft-motion.cpp : never set to true when a new block is loaded (correctly set in current bugfix). Sensorless homing is fine, BLTouch HS mode is fine too. Waiting for feedback

**Comment by thisiskeithb** [comment]
> Also, these machines sound terrible with `FT_MOTION` enabled, so there's probably a lot of re-tuning required compared to standard Input Shaping apparently. 😄

Is it a bug or feature that motors/motions have a high-pitched whine with `FT_MOTION` enabled? I don’t have this problem with standard Input Shaping.

**Comment by thisiskeithb** [comment]
> Is it a bug or feature that motors/motions have a high-pitched whine with `FT_MOTION` enabled? I don’t have this problem with standard Input Shaping.

To answer my own question after re-reading [230119_marlin_fixed_time_ctrl.pdf](https://github.com/MarlinFirmware/Marlin/files/11136945/230119_marlin_fixed_time_ctrl.pdf):

> Known Issues and Miscellaneous
> - The operation noise generated by the machine changes under this control method.

I'll see if I can write up a note for [M493 - Fixed-Time Motion](https://marlinfw.org/docs/gcode/M493.html) since I know I won't be the only one wondering if something was misconfigured.

I don't know about other drivers, but at least for UART & SPI-connected TMCs with various motors (stock Creality, LDO, Prusa - OEM'd by LDO, E3D, other generics), they sound like the current is cranked way too high & slightly crunchy.

**Comment by narno2202** [comment]
@thisiskeithb , on my config, this noise was only present in the early development phase of FT_MOTION, now the sound is quite different but without cracking or anything else. Have you tried to increase the stepper command buffer size? Can you post an audio file?

**Comment by vovodroid** [comment]
I tried [Remove obsolete config parameters](https://github.com/narno2202/Marlin/commit/d0ca905cdf0b2b137c442b3c9f2376654f44ed20) commit

Z homing still fails.

There is indeed some new sound during movements.

Here is recording
[FTM.zip](https://github.com/user-attachments/files/15936232/FTM.zip)

Strong 400 and 200 Hz signal is present with FTM enabled (w/o input shaping)

image

Oscillograms:

No FTM, holding state
image

FTM, holding state
image

Frequency is about 15-20% higher, voltage seems to be the same.


Motion oscillograms (video):
No FTM

https://github.com/MarlinFirmware/Marlin/assets/8691781/0d5a0b1d-444d-495c-9cd8-2d9c9aa009e5

FTM

https://github.com/MarlinFirmware/Marlin/assets/8691781/3bd7470d-3db5-4b7b-901a-653ff4fb231e

**Comment by vovodroid** [comment]
It seems that FMT produces sharper spikes and  tails:

No FMT
image

FMT
image

**Comment by narno2202** [comment]
@vovodroid , could you comment line 3206 in planner.cpp `if (TERN0(FT_MOTION, ftMotion.cfg.active)) synchronize()` in this commit and give feeback us on your Z Axis homing failure.

**Comment by vovodroid** [comment]
Unfortunately nothing changed, even when built with old IS disabled.

Would you like me to try with some logs added?

**Comment by vovodroid** [comment]
I also tried with this line and latest bugfix merged (with https://github.com/MarlinFirmware/Marlin/pull/27179), the same failure.

**Comment by narno2202** [comment]
Thanks @vovodroid . Have you tried to enable ENDSTOP_NOISE_THRESHOLD? I'm waiting for a BIqu Microprobe to do more testing. It should be of interest to have feedback with other probes (my BLTouch and 2 other similar probes work fine).

**Comment by vovodroid** [comment]
> Have you tried to enable ENDSTOP_NOISE_THRESHOLD?

I can't as `#error "SENSORLESS_HOMING is incompatible with ENDSTOP_NOISE_THRESHOLD."`


> I'm waiting for a BIqu Microprobe

What could be different here is that BIqu Microprobe is connected to Z-stop connector, not Z-probe.

**Comment by narno2202** [comment]
The BLTouch probe is also connected to the Z min endstop on my moterboard

**Comment by narno2202** [comment]
Finally iI get a Biqu Microprobe V2. I can't get it working for Z homing in current bugfix without `FT_MOTION` enabled in `Configuration_adv.h` and in this PR with or without `FT_MOTION`. I look at @vovodroid configuration files, I test many configurations but It always fails.

**Comment by vovodroid** [comment]
> I can't get it working for Z homing in current bugfix without `FT_MOTION`

What are symptoms?

Here is relevant parameters:
```
//#define ENDSTOPPULLUPS
#if DISABLED(ENDSTOPPULLUPS)
  // Disable ENDSTOPPULLUPS to set pullups individually
  //#define ENDSTOPPULLUP_XMIN
  //#define ENDSTOPPULLUP_YMIN
  #define ENDSTOPPULLUP_ZMIN
........
#endif

//#define ENDSTOPPULLDOWNS

#define Z_MIN_ENDSTOP_HIT_STATE LOW
#define Z_MIN_PROBE_ENDSTOP_HIT_STATE LOW
#define BIQU_MICROPROBE_V2    // Triggers LOW
#define PROBE_ENABLE_DISABLE
```

I leave XY endstop in floating state, as I have sensorless homing. And `#define Z_MIN_PROBE_ENDSTOP_HIT_STATE LOW` probably doesn't matter, as probe connected to Z-stop connector.

**Comment by narno2202** [comment]
I need to define #define ENDSTOPPULLUP_XMIN, #define ENDSTOPPULLUP_YMIN and MAX for X and Y axis as I use sensorless homing. I use the same configuration as you described. I have the same failure as in your previous posts.

**Comment by vovodroid** [comment]
As it's reproducible I believe  we can catch it with logs.

**Comment by thinkyhead** [comment]
Let's merge this to get motion stuff settled, then address endstop and whining in a new PR.

**Pull request merged**